### ISSUE TITLE
How to customize the number of keypoints while training poseC3d?

### ISSUE THREAD

**Issue by RelayZ** [open-mmlab/mmaction2#1269]
Thanks for your error report and we appreciate it a lot.
If you feel we have help you, give us a STAR! :satisfied:

**Describe the bug**

I'm trying to train poseC3d with my dataset which contrains 13 keypoints only. I altered the `in_channels = 17` to `13` and modified the `left_kp`and `right_kp` in the config file. However, when I started to train it, the `loss` and `loss_cls` is NaN. 

I wonder if I should modified the soure code to make it work, or just simply use another dataset with 17 keypoints.

Thank u!

**Comment by kennymckormick** [comment]
Hey, RelayZ, 
    I guess these modifications are sufficient for poseC3D on your custom dataset. Maybe there is some bug in your data format, please check https://github.com/open-mmlab/mmaction2/tree/master/tools/data/skeleton#the-format-of-posec3d-annotations to make sure your data format is OK.

**Comment by RelayZ** [comment]
Thank you for your reply. It really helped me a lot.

I solved it through debug just a few hours ago for the reason you mentioned. It was simply because some keypoint coordinates in my dataset were empty and I forgot to initialize it when I used pandas.

Another question that comes up. There are some situations in my dataset where the human body temporarily disappears or is blocked in some frames. How do I handle these frames while training my custom dataset? I think there are two approaches:

1.  Delete frames with empty coordinates of keypoints. It's going to change the `T(temporal length)` of the `keypoint` in the annotation. Does that affect the normal training process of poseC3D?

2.  Initialize the keypoints of these frames to a value, such as `0`.

Which method should I choose?

Thank u!

**Comment by kennymckormick** [comment]
Both solutions are OK, I think. You can test which one is better via experiments.

**Comment by bit-scientist** [comment]
Hi, @RelayZ. Could you please elaborate more on what modifications you did with keypoint? I also need to do a similar task. Thank you.

**Comment by RelayZ** [comment]
> Hi, @RelayZ. Could you please elaborate more on what modifications you did with keypoint? I also need to do a similar task. Thank you.



Sry, @bit-scientist. I am sorry that I did not reply your question in time because I was too busy these days. I'm glad to take your questions and also to review the work I've done.

First, you should takes a lot of effort to get the data format right on [custom training1](https://github.com/open-mmlab/mmaction2/tree/master/tools/data/skeleton#the-format-of-posec3d-annotations).And some changes in [custom training2](https://github.com/open-mmlab/mmaction2/blob/master/configs/skeleton/posec3d/custom_dataset_training.md).

Second, I made the following changes to the `slowonly_r50_u48_240e_ntu120_xsub_keypoint.py` file:

1. Because my dataset has only 13 key points

`backbone = dict (   
type='ResNet3dSlowOnly',   
depth=50,   
pretrained=None,
In_channels =13, # It was 17 before`

2.  Because my data set has only 13 key points
`Left_kp = [1, 3, 5, 7, 9, 11,13]
right_kp = [2, 4, 6, 8, 10, 12, ]`

3.  I trained from scratch.So
`Load_from = None` 
`optimizer = dict(
Type = 'SGD, lr = 0.01, momentum = 0.9,
Weight_decay = 0.03)`

Finally, I trained on the RTX 2080Ti.
python `tools/train.py configs/skeleton/posec3d/slowonly_r50_u48_240e_ntu120_xsub_keypoint.py --work-dir  work_dirs/slowonly_r50_u48_240e_ntu120_xsub_keypoint --validate --test-best --gpus 1 --seed 0 --deterministic`

At present, when I initialize the default coordinate to 0 and train 120 epochs, the accuracy rate can reach 91%. I'll test another method in the next few days.

If I can help you, I will be very happy.

**Comment by shelly-xue** [comment]
@RelayZ  Hello, Did you obtain skeleton data according to the steps of custom dataset in mmaction2? I use the alphapose pose estimation algorithm to obtain 17 key points in the form of (x, y, c), but the video is multi person, so the skeleton data of two adjacent frames cannot be guaranteed to be continuous in time. How do you deal with it? For example, tracking algorithm or manual processing method is used?
thank you very much!
